test/python/circuit/test_parameters.py

```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
+        """Referentially equal instructions are treated as same."""
+
+        theta = Parameter("theta")
+        gate = RZGate(theta)
+
+        refs = ParameterReferences(((gate, 0), (gate, 0)))
+
+        self.assertIn((gate, 0), refs)
+        self.assertEqual(len(refs), 1)
+        self.assertIs(next(iter(refs))[0], gate)
+        self.assertEqual(next(iter(refs))[1], 0)
+
+    def test_extend_refs(self):
+        """Extending references handles duplicates."""
+
+        theta = Parameter("theta")
+        ref0 = (RZGate(theta), 0)
+        ref1 = (RZGate(theta), 0)
+        ref2 = (RZGate(theta), 0)
+
+        refs = ParameterReferences((ref0,))
+
+        refs |= ParameterReferences((ref0, ref1, ref2, ref1, ref0))
+
+        self.assertEqual(list(refs), [ref0, ref1, ref2])
+
+    def test_copy_param_refs(self):
+        """Copy of parameter references is a shallow copy."""
+
+        theta = Parameter("theta")
+        ref0 = (RZGate(theta), 0)
+        ref1 = (RZGate(theta), 0)
+        ref2 = (RZGate(theta), 0)
+        ref3 = (RZGate(theta), 0)
+
+        refs = ParameterReferences((ref0, ref1))
+        refs_copy = refs.copy()
+        refs.add(ref2)
+
+        # check copy not modified
+        self.assertNotIn(ref2, refs_copy)
+        self.assertEqual(list(refs_copy), [ref0, ref1])
+
+        refs_copy.add(ref3)
+
+        # check original not modified
+        self.assertNotIn(ref3, refs)
+        self.assertEqual(list(refs), [ref0, ref1, ref2])
+
+
+class TestParameterTable(QiskitTestCase):
+    """Test the ParameterTable class."""
+
+    def test_init_param_table(self):
+        """Parameter table init from mapping."""
+
+        p1 = Parameter("theta")
+        p2 = Parameter("theta")
+
+        ref0 = (RZGate(p1), 0)
+        ref1 = (RZGate(p1), 0)
+        ref2 = (RZGate(p2), 0)
+
+        mapping = {p1: ParameterReferences((ref0, ref1)), p2: ParameterReferences((ref2,))}
+
+        table = ParameterTable(mapping)
+
+        # make sure editing mapping doesn't change `table`
+        del mapping[p1]
+
+        self.assertEqual(list(table[p1]), [ref0, ref1])
+        self.assertEqual(list(table[p2]), [ref2])
+
+    def test_set_references(self):
+        """References replacement by parameter key."""
+
+        p1 = Parameter("theta")
+
+        ref0 = (RZGate(p1), 0)
+        ref1 = (RZGate(p1), 0)
+
+        table = ParameterTable()
+        table[p1] = ParameterReferences((ref0, ref1))
+        self.assertEqual(list(table[p1]), [ref0, ref1])
+
+        table[p1] = ParameterReferences((ref1,))
+        self.assertEqual(list(table[p1]), [ref1])
+
+    def test_set_references_from_iterable(self):
+        """Parameter table init from iterable."""
+
+        p1 = Parameter("theta")
+
+        ref0 = (RZGate(p1), 0)
+        ref1 = (RZGate(p1), 0)
+        ref2 = (RZGate(p1), 0)
+
+        table = ParameterTable(
+            {
+                p1: ParameterReferences((ref0, ref1)),
+            }
+        )
```

Comment:
Nitpicky, but I think this should fit on one line, it's just the trailing comma on line 1907 that puts black in "force individual lines" mode for the dict comprehension.

Reply:
Fixed in 1ff03a5.